### PR TITLE
[QDP] Add CUDA kernel tests for basis encoding

### PR DESCRIPTION
**Pull request by tongyu0924**
### Changes

- [x] Test

### Why

`basis.cu` had no test coverage.

### How

Add `tests/basis_encode.rs` covering:
- Single-sample encode: first, middle, last index (f64 + f32)
- Batch encode: multiple samples with different indices (f64 + f32)
- Error rejection: out-of-range index, zero `state_len`, zero `num_samples`
- Non-Linux dummy stub

## Checklist

- [x] Added or updated unit tests for all changes
- [ ] Added or updated documentation for all changes

### CODE REVIEW

File: qdp/qdp-kernels/tests/basis_encode.rs [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+        Err(_) => {
+            println!("SKIP: No CUDA device available");
+            return;
+        }
+    };
+
+    let num_samples = 4usize;
+    let state_len = 4usize;
+    let num_qubits = 2u32;
+    let basis_indices: Vec<usize> = vec![0, 1, 2, 3];
```
Comment (ryankert01) [inline qdp/qdp-kernels/tests/basis_encode.rs]: Can we have multiple test cases and hard testcase that test the limit and edge case of the code?

**Comment by 400Ping** [comment]
Please fix pre-commit

**Review by 400Ping — CHANGES_REQUESTED** [review]
Thanks for adding kernel-level test coverage for basis encoding. I think this PR should be tightened in a few places:

1. Please reduce the duplicated test setup/assertion logic.
   Most tests repeat CUDA device creation, allocation, launch, copy-back, and basis-state assertions. A small helper such as `cuda_device_or_skip()`, `assert_basis_state_f64()`, `assert_basis_state_f32()`, and table-driven cases for first/middle/last would make this much easier to maintain.

2. Please add or clarify the batch-kernel edge cases.
   The batch kernels rely on `state_len == 2^num_qubits` because they compute:
   - `sample_idx = global_idx >> num_qubits`
   - `element_idx = global_idx & (state_len - 1)`

   Right now the launcher only rejects zero `num_samples` / `state_len`. If the intended contract is that callers must validate `state_len`, `num_qubits`, and basis-index bounds before calling the launcher, please document that in the test or PR description. Otherwise, please add validation/tests for:
   - batch out-of-range basis index
   - `state_len` / `num_qubits` mismatch
   - f32 batch zero `state_len`

3. Please assert the imaginary part in the f32 batch test too.
   `test_basis_encode_batch_f32_basic` currently checks only `actual.x`, while the f64 batch test checks both real and imaginary values.

**Comment by rich7420** [comment]
@tongyu0924 thanks for the patch!

nits: the state buffers are created with `alloc_zeros`, so the "all other elements should be 0" assertions pass even if the kernel never writes those slots — they're zero to begin with. Since basis encoding's contract is to overwrite the *entire* state vector, pre-filling the buffer with a sentinel (e.g. `1.0`) before launch would actually verify the kernel zeroes them. Matters most for the batch kernels, where a grid-stride indexing bug could silently skip elements.

**Comment by viiccwen** [comment]
@tongyu0924 any updates?

**Comment by tongyu0924** [comment]
> @tongyu0924 any updates?

Sorry, I've been pretty busy lately and haven't had much time to work on this. Please feel free to let someone else take over if they'd like to continue with it.
